Client/Frontend/ContentBlocker/ContentBlockerHelper.swift

```diff
@@ -0,0 +1,316 @@
+    }
+
+    class func name() -> String {
+        return "ContentBlockerHelper"
+    }
+
+    private static var heavyInitHasRunOnce = false
+
+    init(tab: Tab, profile: Profile) {
+        self.ruleStore = WKContentRuleListStore.default()
+        if self.ruleStore == nil {
+            print("WKContentRuleListStore unavailable.")
+            assert(false)
+            return
+        }
+
+        self.tab = tab
+        self.profile = profile
+
+        NotificationCenter.default.addObserver(self, selector: #selector(ContentBlockerHelper.reloadTab), name: NSNotification.Name(rawValue: NotificationContentBlockerReloadNeeded), object: nil)
+
+        addActiveRulesToTab(reloadTab: false)
+
+        if ContentBlockerHelper.heavyInitHasRunOnce {
+            return
+        }
+        ContentBlockerHelper.heavyInitHasRunOnce = true
+
+        var wasRuleListChangedDuringInit = false
+        removeOldListsByDateFromStore() { isListChanged in
+            wasRuleListChangedDuringInit = isListChanged
+
+            self.removeOldListsByNameFromStore() { isListChanged in
+                wasRuleListChangedDuringInit = wasRuleListChangedDuringInit || isListChanged
+
+                self.compileListsNotInStore() { isListChanged in
+                    wasRuleListChangedDuringInit = wasRuleListChangedDuringInit || isListChanged
+
+                    if wasRuleListChangedDuringInit {
+                        NotificationCenter.default.post(name: Notification.Name(rawValue: NotificationContentBlockerReloadNeeded), object: nil)
+                    }
+                }
+            }
+        }
+    }
+
+    deinit {
+        NotificationCenter.default.removeObserver(self)
+    }
+
+    fileprivate func readStrengthPref() -> StrengthOption {
+        let pref = profile?.prefs.stringForKey(ContentBlockerHelper.PrefKeyStrength) ?? ""
+        return StrengthOption(rawValue: pref) ?? .basic
+    }
+
+    fileprivate func readEnabledPref() -> EnabledOption {
+        let pref = profile?.prefs.stringForKey(ContentBlockerHelper.PrefKeyEnabledState) ?? ""
+        return EnabledOption(rawValue: pref) ?? .onInPrivateBrowsing
+    }
+
+    @objc func reloadTab() {
+        addActiveRulesToTab(reloadTab: true)
+    }
+
+    fileprivate func addActiveRulesToTab(reloadTab: Bool) {
+        guard let ruleStore = ruleStore else { return }
+        let rules = blocklistBasic + (readStrengthPref() == .strict ? blocklistStrict : [])
+        let enabledMode = readEnabledPref()
+        removeAllFromTab()
+
+        func addRules() {
+            var completed = 0
+            for name in rules {
+                ruleStore.lookUpContentRuleList(forIdentifier: name) { rule, error in
+                    self.addToTab(contentRuleList: rule, error: error)
+                    completed += 1
+                    if reloadTab && completed == rules.count {
+                        self.tab?.reload()
```

Comment:
The suggestion you made elsewhere to delay the tab restoration instead to wait for this code to complete is a good idea.